examples/atomspace/set-numbers.scm

```diff
@@ -0,0 +1,66 @@
+;; Construct Zermelo and Von Neumann numbers
+
+(define (zermelo n)
+"
+  Construct the first n+1 Zermelo numbers
+
+  (zermelo 0) -> (Set)
+  (zermelo n) -> (Set (zermelo n-1))
+"
+  (if (= n 0)
+      (Set)
+      (Set (zermelo (+ n -1)))))
```

Comment:
something wrong with saying `(- n 1)` ?

Reply:
You also realize that the function that you wrote is not tail-recusrsive, and thus it blows out the scheme stack?  So the RAM that you are measuring here includes guile RAM just as much as the atomspace RAM. 

Since tail-recursion tends to be example #2 in "how to program in scheme", we really should not violate that in examples.  Ditto for von neumann, below.

Reply:
I hope so, I mean that's the RAM measured after algorithm has run, thus the memory stack no longer needs to be hold in memory, however I'm not sure Guile disallocates it.

BTW, I tried to save that in a file and load it to be sure it's not the stack but I failed (I think because I tried to write the string at once and Linux didn't let me allocate the space for it). I guess I should write an iterative or tail-recursive version of it, but maybe it's too atypical no bother anyway.

Reply:
Its just that writing recursive functions that are not tail recursive is just bad form, esp. here where it't not any harder.

As to RAM: The operating system reports the location of the `sbrk` -- the top of the heap -- as the as the size of the running process, even if the heap is almost entirely unallocated. By allocating lots of things in scheme, you force the heap to grow.  Even if they are freed later on, the increase of the size of the heap is permanent: the heap never shrinks.

Reply:
The one-way growth of the heap is not a guile thing, its a glibc thing.  I think I once saw some code in guile that tried to return excess guile-heap to glibc, but its not clear that glibc can actually shrink sbrk as a result (due to memory fragmentation -- just like disk fragmentation in MSDOS).  I think glibc can in some cases unmap virtual empty pages so that there is no physical page behind it; the operating system only counts physical pages in the RSS.  All of these various optimization tricks make it hard to predict what is actually happening. It's best to not allocate the memory in the first place, if you don't actually need it.

Reply:
I thought that sort of recursive calls would be allocated on the stack mostly, not the heap. However after reading a bunch on the subject, it seems the stack doesn't get dynamical disallocated by the OS either, so same problem.